components/eamxx/src/physics/zm/eamxx_zm_process_interface.cpp

```diff
@@ -171,123 +194,114 @@ void ZMDeepConvection::run_impl (const double dt)
-        auto thl_std_pbl  = sqrt( thl_sec_pbl ); // std deviation of thetal;
-        zm_input.tpert(i) = ( thl_std_pbl + (latvap/cpair)*qc_pbl ) / exner_pbl;
-        zm_input.tpert(i) = std::min(2.0,zm_input.tpert(i)); // apply limiter
-      }
-    }
-  });
+  zm_input.calculate_tpert(m_ncol,m_nlev,is_first_step);
 
   //----------------------------------------------------------------------------
   // run the ZM scheme
 
   zm_eamxx_bridge_run(m_ncol, m_nlev, zm_input, zm_output, zm_opts);
 
   //----------------------------------------------------------------------------
-  // update prognostic fields
+  // create temporaries of output variables to avoid "Implicit capture" warning
 
-  const auto zm_output_loc = zm_output; // create temp to avoid "Implicit capture" warning
+  const auto loc_zm_output_prec    = zm_output.prec;
+  const auto loc_zm_output_snow    = zm_output.snow;
+  const auto loc_zm_output_cape        = zm_output.cape;
+  const auto loc_zm_output_activity    = zm_output.activity;
+  const auto loc_zm_output_tend_t  = zm_output.tend_t;
+  const auto loc_zm_output_tend_qv = zm_output.tend_qv;
+  const auto loc_zm_output_tend_u  = zm_output.tend_u;
+  const auto loc_zm_output_tend_v  = zm_output.tend_v;
+
+  //----------------------------------------------------------------------------
+  // update prognostic fields
 
   if (zm_opts.apply_tendencies) {
     // accumulate surface precipitation fluxes
     Kokkos::parallel_for("zm_update_precip",KT::RangePolicy(0, m_ncol), KOKKOS_LAMBDA (const int i) {
-      auto prec_liq = zm_output_loc.prec(i) - zm_output_loc.snow(i);
-      precip_liq_surf_mass(i) += std::max(prec_liq,0.0) * PC::RHO_H2O * dt;
-      precip_ice_surf_mass(i) += zm_output_loc.snow(i)  * PC::RHO_H2O * dt;
+      Real prec_liq = loc_zm_output_prec(i) - loc_zm_output_snow(i);
+      precip_liq_surf_mass(i) += ekat::impl::max(0.0,prec_liq) * PC::RHO_H2O * dt;
+      precip_ice_surf_mass(i) += loc_zm_output_snow(i) * PC::RHO_H2O * dt;
     });
-    // update 3D prognostic variables
-    Kokkos::parallel_for("zm_update_prognostics",team_policy, KOKKOS_LAMBDA (const KT::MemberType& team) {
-      const int i = team.league_rank();
-      Kokkos::parallel_for(Kokkos::TeamVectorRange(team, nlevm_packs), [&](const int k) {
-        T_mid(i,k) += zm_output_loc.tend_s (i,k)/cpair * dt;
-        qv   (i,k) += zm_output_loc.tend_qv(i,k)       * dt;
-        uwind(i,k) += zm_output_loc.tend_u (i,k)       * dt;
-        vwind(i,k) += zm_output_loc.tend_v (i,k)       * dt;
-      });
+
+    Kokkos::parallel_for("zm_update_precip",KT::RangePolicy(0, m_ncol*nlev_mid_packs), KOKKOS_LAMBDA (const int idx) {
+      const int i = idx/nlev_mid_packs;
+      const int k = idx%nlev_mid_packs;
+      T_mid(i,k) += loc_zm_output_tend_t (i,k) * dt;
+      qv   (i,k) += loc_zm_output_tend_qv(i,k) * dt;
+      uwind(i,k) += loc_zm_output_tend_u (i,k) * dt;
+      vwind(i,k) += loc_zm_output_tend_v (i,k) * dt;
```

Comment:
Why not nested team policy here? Not a concern either way since it's in initialization, just curious.

Reply:
I find MDRange much cleaner and easier to understand. I know Jim found some cases in rrtmgp where MDRange was slower than a "manual mdrange" (the stuff you are doing here), but the kokkos devs claim this should not happen, so I think it must be some sort of peculiar rrtmgp case. Here, I would vote for clarity over the few micro seconds we may gain in perf...